Enable nxp mcus general power off demo

ZhaoxiangJin opened the pull request:
1. Enabled generic poweroff example for NXP MCUs. The example can test pass on frdm_mcxa153, frdm_mcxa156, frdm_mcxa166, frdm_mcxa276, frdm_mcxn236, frdm_mcxn947 and mimxrt595_evk.
2. Migrate the existing mimxrt595_evk/system_off example to the new poweroff example.

ZhaoxiangJin committed:
modules: hal_nxp: Pulling cmc, vbat, wuu driver

Pulling MCUX SDK cmc, vbat, wuu driver to Zephyr build tree.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards/nxp: mcxn: Enable LPUART pin internal pullup resistor

This commit enables MCXN236, MCXN947 LPUART pin internal
pullup resistor. For MCXN947 and MCXN236, during LPUAR
initialization, the RX pin is pulled down internally
and STAT[RAF] is set to one. In this state, attempting
to enter low power mode will trigger LPACK reset and
therefore cannot truly enter low power mode. The correct
setting should be to enable LPUART pin internal pullup
resistor.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

dleach02 commented:
@ZhaoxiangJin please address the CI issues

ZhaoxiangJin commented on samples/boards/nxp/poweroff/src/main.c:
```diff
@@ -0,0 +1,85 @@
+#include <zephyr/drivers/gpio.h>
+#endif
+
+int main(void)
```
will splitting those different codes into separate files

ZhaoxiangJin commented on samples/boards/nxp/poweroff/Kconfig:
```diff
@@ -0,0 +1,22 @@
+
+mainmenu "NXP MCU POWER OFF DEMO"
+
+config COUNTER_WAKEUP_ENABLE
+	bool
+	select COUNTER
+	default y if BOARD_FRDM_MCXA153 || BOARD_FRDM_MCXA156 || \
+				BOARD_FRDM_MCXA166 || BOARD_FRDM_MCXA276 || \
+				BOARD_FRDM_MCXN947 || BOARD_MIMXRT595_EVK
+	help
+	  Use counter to wake up device from poweroff.
+
+config GPIO_WAKEUP_ENABLE
+	bool
+	select GPIO
+	default y if BOARD_FRDM_MCXN236
+	help
+	  Use gpio to wake up device from poweroff.
```
what's the advantage of using DT to drive this? The advantage of using Kconfig is that users can select the wakeup source (wakeup button or counter) as needed.

ZhaoxiangJin commented:
Ping @decsny for review

decsny commented on soc/nxp/mcx/mcxa/Kconfig.soc:
```diff
@@ -90,3 +90,11 @@ config SOC_PART_NUMBER
 	default "MCXA276VLL" if SOC_PART_NUMBER_MCXA276VLL
 	default "MCXA276VLH" if SOC_PART_NUMBER_MCXA276VLH
 	default "MCXA276VPN" if SOC_PART_NUMBER_MCXA276VPN
+
+config SOC_WAKEUP_DELAY
+	int "Wakeup delay"
+	default 91
+	help
+	  When a different VDD_CORE level is configured in low
+	  power modes compared to active modes, the wakeup delay
+	  must be configured to a minimum value.
```
this config should not be added in a Kconfig.soc file, Kconfig.soc is only for HWM hierarchy description, put in Kconfig file and namespace for the platform

decsny commented on soc/nxp/mcx/mcxn/Kconfig.soc:
```diff
@@ -65,3 +65,11 @@ config SOC_PART_NUMBER
 	default "MCXN236VDF" if SOC_PART_NUMBER_MCXN236VDF
 	default "MCXN236VNL" if SOC_PART_NUMBER_MCXN236VNL
+
+config SOC_WAKEUP_DELAY
+	int "Wakeup delay"
+	default 780
+	help
+	  When a different VDD_CORE level is configured in low
+	  power modes compared to active modes, the wakeup delay
+	  must be configured to a minimum value.
```
ditto

decsny commented on soc/nxp/mcx/mcxn/power.c:
the power.c code looks almost the same between mcxa and mcxn, not required but would it be a good idea or not to share code somewhere in a file somewhere like mcx/common ?

decsny reviewed (changes requested):
blocked over kconfig namespaces and Kconfig.soc

ZhaoxiangJin commented:
Update:
1. Renamed kconfig option from 'SOC_WAKEUP_DELAY' to 'SOC_MCXN_WAKEUP_DELAY' and moved it from Kconfig.soc to Kconfig
2. Changed the wakeup source Kconfig option in the poweroff demo Kconfig choice config, because the poweroff example currently only supports a single wakeup source. In the future, if there is demand, the example can be further expanded to support two wakeup sources at the same time.

ZhaoxiangJin commented:
Update:
rebased to fix ci issue

ZhaoxiangJin commented:
Update:
Fixed conflicts

decsny commented on drivers/counter/Kconfig.mcux_lptmr:
```diff
@@ -3,9 +3,22 @@
-	depends on DT_HAS_NXP_LPTMR_ENABLED
-	help
-	  Enable support for the MCUX Low Power Timer (LPTMR).
+if MCUX_HW_HAS_LPTMR0
+mcux_hw_lptmr_index = 0
+rsource "Kconfig.mcux_lptmr_instance"
+endif
+
+if MCUX_HW_HAS_LPTMR1
+mcux_hw_lptmr_index = 1
+rsource "Kconfig.mcux_lptmr_instance"
+endif
+
+if MCUX_HW_HAS_LPTMR2
+mcux_hw_lptmr_index = 2
+rsource "Kconfig.mcux_lptmr_instance"
+endif
+
+if MCUX_HW_HAS_LPTMR3
+mcux_hw_lptmr_index = 3
+rsource "Kconfig.mcux_lptmr_instance"
+endif
```
what is going on here? this is completely against the design philosophy of the project

ZhaoxiangJin replied:
At the Kconfig level, we need to determine whether each LPTMR instance has the wakeup-source function, and then define 'COUNTER_MCUX_LPTMR_WAKEUP_SOURCE_INDEX'.

decsny replied:
no, this is what an alias is for.

decsny commented on soc/nxp/common/Kconfig.peripherals:
And not to mention I just had a big PR removing all these legacy HAS_MCUX configs, and im not about to allow a new paradigm that starts adding them back again.


Looking at the commit that adds this in isolation, it seems completely pointless, but I will keep looking at the next commits for the context.

decsny commented on dts/bindings/gpio/nxp,kinetis-gpio.yaml:
```diff
@@ -24,6 +24,26 @@ properties:
     description: |
       Describes an offset between inst index and actual GPIO port number.
 
+  wakeup-line:
+    type: array
+    description: |
+      On some NXP platforms, GPIO can wake up the MCU, but the wakeup capability
+      is configured through a specific peripheral, such as the wakeup unit. We need
+      to use this property to specify which pin has the wakeup capability and which
+      register of the wakeup unit the pin is mapped to.
+      For example, configure wakeup-line = <20 4> indicates GPIOn_20 can wakeup
+      the SoC/MCU, and we need to configure wakeup unit register 4 to enable wakeup
+      function.
+
+  wakeup-signal-edge:
+    type: int
+    enum:
+      - 1 # External input Pin enabled with the rising edge detection
+      - 2 # External input Pin enabled with the falling edge detection.
+      - 3 # External input Pin enabled with any change detection.
+    description: |
+      Indicates which edge signal on the wakeup pin can wake up the MCU.
+
```
i dont think these belong as properties, they should be cells in a specifier for something else, issue is that there is no standard specifier space , which is why im adding mux as one , but i also dont want to overuse mux

decsny commented on samples/boards/nxp/poweroff/boards/frdm_mcxa156.overlay:
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright 2025 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/ {
+	aliases {
+		poweroff-wakeup-source = &lptmr0;
+	};
+};
+
+&lptmr0 {
+	status = "okay";
+	timer-mode-sel = <0>;
+	clk-source = <1>;
+	clock-frequency = <16384>;
+	resolution = <32>;
+	wakeup-source;
+};
```
all these MCXA board overlay for this sample are the same, @hakehuang what is your opinion about how to do this one ? make one overlay for all of them? or it complicates the testcase yaml ? (but we already are adding a lot of cases anyways)

decsny commented on doc/_scripts/redirects.py:
```diff
@@ -253,7 +253,7 @@
     ('samples/boards/mimxrt1060_evk/system_off/README', 'samples/boards/nxp/mimxrt1060_evk/system_off/README'),
     ('samples/boards/mimxrt1170_evk_cm7/magic_addr/README', 'samples/boards/nxp/mimxrt1170_evk_cm7/magic_addr/README'),
-    ('samples/boards/mimxrt595_evk/system_off/README', 'samples/boards/nxp/mimxrt595_evk/system_off/README'),
+    ('samples/boards/mimxrt595_evk/system_off/README', 'samples/boards/nxp/poweroff/README'),
```
this commit should be squashed to the previous commit that actually did the migrating, not a separate commit

ZhaoxiangJin replied:
squashed

decsny reviewed (changes requested):
I still don't understand what is the point of having the HAS_MCUX_LPTMR_INST_0 etc macros, this really needs to be avoided

ZhaoxiangJin commented on soc/nxp/mcx/mcxn/soc.c:
```diff
@@ -16,11 +16,23 @@
 
 void soc_reset_hook(void)
 {
+#if defined(CONFIG_PM) || defined(CONFIG_POWEROFF)
+	/* Release the I/O pads and certain peripherals to normal run mode state,
+	 * for in Power Down mode they will be in a latched state.
+	 */
+	if ((CMC_GetSystemResetStatus(CMC0) & kCMC_WakeUpReset) != 0UL) {
+		SPC_ClearPeriphIOIsolationFlag(SPC0);
+	}
+#endif
```
the operation 'SPC_ClearPeriphIOIsolationFlag(SPC0);' should placed under soc_reset_hook before SystemInit.

ZhaoxiangJin commented on boards/nxp/frdm_mcxa346/board.c:
```diff
@@ -268,6 +271,9 @@ void board_early_init_hook(void)
 	CLOCK_SetupFRO16KClocking(kCLKE_16K_SYSTEM | kCLKE_16K_COREMAIN);
+#if defined(CONFIG_PM) || defined(CONFIG_POWEROFF)
+	VBAT_EnableFRO16k(VBAT0, true);
+#endif
```
When a PR attempts to add large amounts of duplicate code, the reviewer should be alerted.

ZhaoxiangJin committed:
soc: nxp: rt5xx: Improve poweroff for rt5xx

This commit introduced the following changes:
1. MIMXRT595 poweroff corresponds to its full
deep power down mode not deep power down mode.
2. Enable OSC32K before poweroff, the OSC32K
is the clock source of the wakeup source RTC.
3. Enable RTC wake-up function.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
soc: nxp: mcxa/n: Enable poweroff for mcxa/n series

This commit introduced the following changes:
1. Selected kconfig option 'HAS_POWEROFF' for NXP
   MCXA/N series.
2. Added power.c for NXP MCXA/N series, in this file,
   we currently implemented 'z_sys_poweroff()' function.
3. For the MCXA/N series, after waking up from the deep
   power down mode, the reset handler will be executed,
   and we need to release the I/O pads and certain peripheral
   devices to normal operating mode in 'soc_reset_hook'.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp: mcxa/n: Enable FRO16k for LPTMR

Enabled FRO16k for LPTMR when LPTMR clock source
is set to 0x1.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: counter/lptmr: Enable lptmr wakeup feature

1. Enable lptmr wakeup feature
2. Enable lptmr get_freq function

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: gpio: Enable gpio_mcux wakeup feature

Enable gpio_mcux wakeup feature

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: bindings: Added nxp,spc.yaml

dts: bindings: Added nxp,spc.yaml

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: nxp: Added spc for mcxa/n

1. Added SPC for mcxa/n
2. Added new property 'wakeup-configure-reg-idx'
for lptmr

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
samples: boards/nxp: Enable generic poweroff for NXP MCUs

This commit introduced the following changes:
1. Enabled generic poweroff example for NXP MCUs. The
example is test on frdm_mcxa153, frdm_mcxa156, frdm_mcxa166,
frdm_mcxa276, frdm_mcxn236, frdm_mcxn947 and mimxrt595_evk.

2. Migrate the existing mimxrt595_evk/system_off example
to the new poweroff example.

3. The Poweroff example now supports counter wake-up and
button/GPIO wake-up, suitable for most NXP devices/boards.
For testing purposes, frdm_mcxn236 is wakeup using the
wake-up button; frdm_mcxa153, frdm_mcxa156, frdm_mcxa166,
frdm_mcxa276, and frdm_mcxn947 are wakeup using lptmr;
mimxrt595_evk is wakeup using high-resolution RTC.

4. Update redirects.py to find the correct mimxrt595_evk
system_off example README.

5. Update release-notes to record the new nxp poweroff
example.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -17,6 +17,13 @@
+	defined(CONFIG_SOC_FAMILY_MCXW)
+#include <fsl_wuu.h>
+#endif
+#endif
```
removed this code, use '#include <soc.h>' instead.

ZhaoxiangJin commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -405,6 +443,21 @@ static void gpio_mcux_port_isr(const struct device *dev)
 
+#if defined(CONFIG_PM) || defined(CONFIG_POWEROFF)
+#if defined(CONFIG_SOC_FAMILY_MCXN) || defined(CONFIG_SOC_FAMILY_MCXA) || \
+	defined(CONFIG_SOC_FAMILY_MCXW)
```
removed `#if defined(CONFIG_SOC_FAMILY_MCXN) || defined(CONFIG_SOC_FAMILY_MCXA) || defined(CONFIG_SOC_FAMILY_MCXW)`. use `#if defined(FSL_FEATURE_SOC_WUU_COUNT)` instead.

ZhaoxiangJin commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -423,18 +476,18 @@ static void gpio_mcux_shared_cluster_isr(const struct device *ports[])
+		irq_enable(DT_IRQN(node_id));							\
+												\
+		return 0;									\
+	}											\
```
removed this change

ZhaoxiangJin commented:
Update:
1.rebased to fix CI issues.

dleach02 commented:
@ZhaoxiangJin can you rebase and resolve the merge conflicts
